### ISSUE TITLE
ParseError: Code(0) with message "syntax error, unexpected ';', expecting :: (T_PAAMAYIM_NEKUDOTAYIM)

### ISSUE THREAD

**Issue by Isinlor** [kahlan/kahlan#363]
Hi!

I'm currently updating my kahlan from version 3 to 4 or 5, since I want it to get to work on PHP 7.4. 

When using kahlan^3.0 everything works correctly.

But when I update to either ^4.0 or ^5.0 I get following error message:

```
      an uncaught exception has been thrown in `src/Constraint/Hard/BlockedRelationshipsConstraint.php` line 95
      
      message:`ParseError` Code(0) with message "syntax error, unexpected ';', expecting :: (T_PAAMAYIM_NEKUDOTAYIM)"
      
        [NA] - src/Constraint/Hard/BlockedRelationshipsConstraint.php, line  to 95
        Kahlan\Jit\ClassLoader::loadFile() - vendor/kahlan/kahlan/src/Jit/ClassLoader.php, line 759
        Kahlan\Jit\ClassLoader::loadClass() - [internal], line ?
        spl_autoload_call - spec/Constraint/Hard/BlockedRelationshipsConstraintSpec.php, line 75 to 76
        Kahlan\Given::__invoke() - vendor/kahlan/kahlan/src/Scope.php, line 103
        Kahlan\Scope::__get() - vendor/kahlan/kahlan/src/Scope.php, line 107
        Kahlan\Scope::__get() - vendor/kahlan/kahlan/src/Scope.php, line 107
        Kahlan\Scope::__get() - spec/Constraint/Hard/BlockedRelationshipsConstraintSpec.php, line 88

```

The relevant lines from my code are:

```php
            // src/Constraint/Hard/BlockedRelationshipsConstraint.php, line  to 95
            $blockedCounter =& DeepArray::getReference($this->blockedCounters, $blockedRelationship, 0);
```

```php
    // spec/Constraint/Hard/BlockedRelationshipsConstraintSpec.php, line 88
                expect($this->constraint)->toBeAnInstanceOf(BlockedRelationshipsConstraint::class);

```

```php
    // spec/Constraint/Hard/BlockedRelationshipsConstraintSpec.php, line 75 to 82
    given('constraint', function () {
        return new BlockedRelationshipsConstraint(
            new ScopedRelationshipsBlocker(
                $this->groupedByElementTypes,
                new CompositeRelationshipsBlocker(...$this->relationshipsBlockers)
            )
        );
    });
```

I tough I will provide you with a nice reproduction of the bug, however I fail at isolating.

Here you can see my attempt: https://github.com/Isinlor/kahlan-issue

Any idea what it may be?

One thing I've noticed that stack trance does not report starting line in `src/Constraint/Hard/BlockedRelationshipsConstraint.php, line  to 95`, not sure if it's relevant.

**Comment by jails** [comment]
did you tried to clear caches `vendor/bin/kahlan -cc` ?

**Comment by Isinlor** [comment]
Thanks for quick response. And yes, I did.

**Comment by jails** [comment]
Can you dump the file in your cache ? I mean the file in `sys_get_temp_dir()/<ProjectName>/src/Constraint/Hard/BlockedRelationshipsConstraint.php`, 15-20 lines around the line 95 should be enough.

**Comment by Isinlor** [comment]
Here is the cache - seems like an operator precedence issue?
```php
    public function addSolution(array $solution)
    {$__KMONKEY__220=\Kahlan\Plugin\Monkey::patched(__NAMESPACE__ , null, 'ksort', $__KMONKEY__220__);$__KMONKEY__221=\Kahlan\Plugin\Monkey::patched(null, 'KUL\BioMed\Distribution\Helper\DeepArray', null, $__KMONKEY__221__);$__KPOINTCUT_ARGS__ = func_get_args(); $__KPOINTCUT_SELF__ = isset($this) ? $this : get_called_class(); if ($__KPOINTCUT__ = \Kahlan\Plugin\Pointcut::before(__METHOD__, $__KPOINTCUT_SELF__, $__KPOINTCUT_ARGS__)) { $r = $__KPOINTCUT__($__KPOINTCUT_ARGS__, $__KPOINTCUT_SELF__); return $r; }

        $blockedRelationships = $this->relationshipsBlocker->getBlockedRelationships($solution);
        foreach ($blockedRelationships as $blockedRelationship) {

            $__KMONKEY__220($blockedRelationship);

            $blockedCounter =&($__KMONKEY__221__?$__KMONKEY__221__: $__KMONKEY__221::getReference($this->blockedCounters, $blockedRelationship, 0));
            $blockedCounter++;

        }

    }
    ```

**Comment by Isinlor** [comment]
From [PHP doc](https://www.php.net/manual/en/language.operators.comparison.php#language.operators.comparison.ternary):
>  Note: Please note that the ternary operator is an expression, and that **it doesn't evaluate to a variable, but to the result of an expression**. This is important to know if you want to return a variable by reference. The statement return $var == 42 ? $a : $b; in a return-by-reference function will therefore not work and a warning is issued. 

And since there can not be reference to result of an expression.

Seems like this could work `$__KMONKEY__221__ ? ($blockedCounter = &$__KMONKEY__221__) : ($blockedCounter = &$__KMONKEY__221__::getReference());`

**Comment by jails** [comment]
Sorry for the delay, I think this issue is fixed in the 5.0.3 version.
